src/helpers/recompute_zones.ts

```diff
@@ -0,0 +1,398 @@
+  // Case we want to remove a zone from the profile: it's the opposite of the previous case
+  const bottomSuccIndex = binarySuccessorSearch(profile, bottomValue, 0, false);
+  if ((bottomSuccIndex % 2 === 0 && !toRemove) || (bottomSuccIndex % 2 !== 0 && toRemove)) {
+    newPoints.push(bottomValue);
+  }
+
+  // add the top and bottom value to the profile and
+  // remove all information between the top and bottom index
+  profile.splice(topPredIndex + 1, bottomSuccIndex - topPredIndex - 1, ...newPoints);
+}
+
+function removeContiguousProfiles(
+  profilesStartingPosition: number[],
+  profiles: Map<number, number[]>,
+  leftIndex: number,
+  rightIndex: number
+) {
+  for (
+    let i = rightIndex === profilesStartingPosition.length - 1 ? rightIndex : rightIndex + 1;
+    i > (leftIndex - 1 === -1 ? 0 : leftIndex - 1);
+    i--
+  ) {
+    if (
+      deepEqualsArray(
+        profiles.get(profilesStartingPosition[i])!,
+        profiles.get(profilesStartingPosition[i - 1])!
+      )
+    ) {
+      profiles.delete(profilesStartingPosition[i]);
+      profilesStartingPosition.splice(i, 1);
+    }
+  }
+}
+
+function constructZonesFromProfiles(
+  profilesStartingPosition: number[],
+  profiles: Map<number, number[]>
+): UnboundedZone[] {
+  const mergedZone: UnboundedZone[] = [];
+  let stackZones: UnboundedZone[] = [];
+  for (let colIndex = 0; colIndex < profilesStartingPosition.length; colIndex++) {
+    const left = profilesStartingPosition[colIndex];
+    const profile = profiles.get(left)!;
+    if (!profile || profile.length === 0) {
+      mergedZone.push(...stackZones);
+      stackZones = [];
+      continue;
+    }
+
+    let right = profilesStartingPosition[colIndex + 1];
+    if (right !== undefined) {
+      right--;
+    }
+
+    const colZones: UnboundedZone[] = [];
+    for (let i = 0; i < profile.length; i += 2) {
+      const top = profile[i];
+      let bottom = profile[i + 1];
+      if (bottom !== undefined) {
+        bottom--;
+      }
+      const colZone: UnboundedZone = {
+        top,
+        left,
+        bottom,
+        right,
+        hasHeader: (bottom === undefined && top !== 0) || (right === undefined && left !== 0),
+      };
+
+      let findCorrespondingZone = false;
+      for (let j = stackZones.length - 1; j >= 0; j--) {
+        const stackZone = stackZones[j];
+        if (stackZone.top === colZone.top && stackZone.bottom === colZone.bottom) {
+          stackZone.right = colZone.right;
+          stackZones.splice(j, 1);
+          colZones.push(stackZone);
+          findCorrespondingZone = true;
+          continue;
+        }
+      }
+      if (!findCorrespondingZone) {
+        colZones.push(colZone);
+      }
+    }
+
+    mergedZone.push(...stackZones);
+    stackZones = colZones;
+  }
+  mergedZone.push(...stackZones);
+  return mergedZone;
+}
+
+function binaryPredecessorSearch(arr: number[], val: number, start = 0, matchEqual = true) {
```

Comment:
Can't you use the `dichotomicSearch` helper that aready exists ?

Reply:
not really, there are cases here where we don't want to match equality even if it exists. See it as "strictly superior" or "strictly inferior".